File: src/lattice.c
#include "lattice.h"

void make_lattice(int Lx, int Ly, Lattice *latt) 
{
  int i, j, n, n_, b, p;

  latt->Lx = Lx;
  latt->Ly = Ly;
  latt->N = Lx * Ly;
  latt->Ns = 2 * latt->N;
  latt->z = 3;
  latt->Nb = latt->z * latt->N;

  latt->a_1[0] = 0.5; 
  latt->a_1[1] = 0.5 * sqrt(3.0);
  latt->a_2[0] = - 0.5;
  latt->a_2[1] = 0.5 * sqrt(3.0);

  latt->b_1[0] = 2.0 * M_PI * latt->a_2[1] / (latt->Lx * (latt->a_1[0] * latt->a_2[1] - latt->a_1[1] * latt->a_2[0])); 
  latt->b_1[1] = - 2.0 * M_PI * latt->a_2[0] / (latt->Lx * (latt->a_1[0] * latt->a_2[1] - latt->a_1[1] * latt->a_2[0])); 
  latt->b_2[0] = - 2.0 * M_PI * latt->a_1[1] / (latt->Ly * (latt->a_1[0] * latt->a_2[1] - latt->a_1[1] * latt->a_2[0])); 
  latt->b_2[1] = 2.0 * M_PI * latt->a_1[0] / (latt->Ly * (latt->a_1[0] * latt->a_2[1] - latt->a_1[1] * latt->a_2[0])); 

  latt->p_a[0] = 0.0;
  latt->p_a[1] = - 0.5;
  latt->p_b[0] = 0.0;
  latt->p_b[1] = 0.5;

  /* Real-space lattice */
  latt->r = (int **) malloc(latt->N * sizeof(int *));
  for (i = 0; i < latt->N; i++) latt->r[i] = (int *) malloc(2 * sizeof(int));

  latt->inv_r = (int **) malloc(latt->Lx * sizeof(int *));
  for (i = 0; i < latt->Lx; i++) latt->inv_r[i] = (int *) malloc(latt->Ly * sizeof(int));

  latt->r_ij = (int ***) malloc(latt->N * sizeof(int **));
  for (i = 0; i < latt->N; i++) { 
    latt->r_ij[i] = (int **) malloc(latt->N * sizeof(int *));
    for (j = 0; j < latt->N; j++) {
      latt->r_ij[i][j] = (int *) malloc(2 * sizeof(int));
    }
  }

  n = 0;
  for (j = 0; j < Ly; j++) {
    for (i = 0; i < Lx; i++) {
      latt->r[n][0] = i;
      latt->r[n][1] = j;
      latt->inv_r[i][j] = n;
      n++;
    }
  }

  for (i = 0; i < latt->N; i++) {
    for (j = 0; j < latt->N; j++) {
      latt->r_ij[i][j][0] = latt->r[i][0] - latt->r[j][0];
      latt->r_ij[i][j][1] = latt->r[i][1] - latt->r[j][1];
    }
  }

  /* Reciprocal lattice */
  latt->k = (int **) malloc(latt->N * sizeof(int *));
  for (i = 0; i < latt->N; i++) latt->k[i] = (int *) malloc(2 * sizeof(int));

  latt->inv_k = (int **) malloc(latt->Lx * sizeof(int *));
  for (i = 0; i < latt->Lx; i++) latt->inv_k[i] = (int *) malloc(latt->Ly * sizeof(int));

  latt->k_ij = (int ***) malloc(latt->N * sizeof(int **));
  for (i = 0; i < latt->N; i++) { 
    latt->k_ij[i] = (int **) malloc(latt->N * sizeof(int *));
    for (j = 0; j < latt->N; j++) {
      latt->k_ij[i][j] = (int *) malloc(2 * sizeof(int));
    }
  }

  n = 0;
  for (j = 0; j < Ly; j++) {
    for (i = 0; i < Lx; i++) {
      latt->k[n][0] = i;
      latt->k[n][1] = j;
      latt->inv_k[i][j] = n;
      n++;
    }
  }

  for (i = 0; i < latt->N; i++) {
    for (j = 0; j < latt->N; j++) {
      latt->k_ij[i][j][0] = latt->k[i][0] - latt->k[j][0];
      latt->k_ij[i][j][1] = latt->k[i][1] - latt->k[j][1];
    }
  }

  /* Sites for lattice and bond list */
  latt->r_sites = (int *) malloc(latt->Ns * sizeof(int));
  latt->sublattice = (int *) malloc(latt->Ns * sizeof(int));

  n = 0;
  for (i = 0; i < latt->Lx; i++) {
    for (j = 0; j < latt->Ly; j++) {
      n_ = latt->inv_r[i][j];

      latt->r_sites[n] = n_;
      latt->sublattice[n] = 0;
      n++;

      latt->r_sites[n] = n_;
      latt->sublattice[n] = 1;
      n++;
    }
  }

  latt->bond_list = (int **) malloc(latt->Nb * sizeof(int *));
  for (i = 0; i < latt->Nb; i++) {
    latt->bond_list[i] = (int *) malloc(2 * sizeof(int));
  }

  latt->bond_type = (int *) malloc(latt->Nb * sizeof(int));
  latt->z_bonds = (int *) malloc(latt->N * sizeof(int));
  latt->y_bonds = (int *) malloc(latt->N * sizeof(int));
  latt->x_bonds = (int *) malloc(latt->N * sizeof(int));

  b = 0;
  p = 0;
  for (i = 0; i < latt->Lx; i++) {
    for (j = 0; j < latt->Ly; j++) {
      n = 2 * latt->inv_r[i][j];

      n_ = 2 * latt->inv_r[i][j] + 1;
      latt->bond_list[b][0] = n;
      latt->bond_list[b][1] = n_;
      latt->bond_type[b] = 2;
      latt->z_bonds[p] = b;
      b++;

      n_ = 2 * latt->inv_r[i][(j + 1) % latt->Ly] + 1;
      latt->bond_list[b][0] = n;
      latt->bond_list[b][1] = n_;
      latt->bond_type[b] = 1;
      latt->y_bonds[p] = b;
      b++;

      n_ = 2 * latt->inv_r[(i + 1) % latt->Lx][j] + 1;
      latt->bond_list[b][0] = n;
      latt->bond_list[b][1] = n_;
      latt->bond_type[b] = 0;
      latt->x_bonds[p] = b;
      b++;
      p++;
    }
  }

  /* Plaquettes */  
  latt->Np = latt->N;
  latt->size_p = 6;
  latt->plaquette_bonds = (int **) malloc(latt->Np * sizeof(int *));
  latt->plaquette_sites = (int **) malloc(latt->Np * sizeof(int *));
  for (p = 0; p < latt->Np; p++) {
    latt->plaquette_bonds[p] = (int *) malloc(latt->size_p * sizeof(int));
    latt->plaquette_sites[p] = (int *) malloc(latt->size_p * sizeof(int));
  }

  for (p = 0; p < latt->Np; p++) {
    latt->plaquette_sites[p][0] = latt->bond_list[latt->z_bonds[p]][0];
    latt->plaquette_sites[p][1] = latt->bond_list[latt->z_bonds[p]][1];
    latt->plaquette_bonds[p][0] = latt->z_bonds[p];

    for (i = 0; i < latt->N; i++) {
      if (latt->bond_list[latt->x_bonds[i]][1] == latt->plaquette_sites[p][1]) {
        latt->plaquette_sites[p][2] = latt->bond_list[latt->x_bonds[i]][0];
        latt->plaquette_bonds[p][1] = latt->x_bonds[i];
        break;
      }
    }

    for (i = 0; i < latt->N; i++) {
      if (latt->bond_list[latt->y_bonds[i]][0] == latt->plaquette_sites[p][2]) {
        latt->plaquette_sites[p][3] = latt->bond_list[latt->y_bonds[i]][1];
        latt->plaquette_bonds[p][2] = latt->y_bonds[i];
        break;
      }
    }

    for (i = 0; i < latt->N; i++) {
      if (latt->bond_list[latt->z_bonds[i]][1] == latt->plaquette_sites[p][3]) {
        latt->plaquette_sites[p][4] = latt->bond_list[latt->z_bonds[i]][0];
        latt->plaquette_bonds[p][3] = latt->z_bonds[i];
        break;
      }
    }

    for (i = 0; i < latt->N; i++) {
      if (latt->bond_list[latt->x_bonds[i]][0] == latt->plaquette_sites[p][4]) {
        latt->plaquette_sites[p][5] = latt->bond_list[latt->x_bonds[i]][1];
        latt->plaquette_bonds[p][4] = latt->x_bonds[i];
        break;
      }
    }

    for (i = 0; i < latt->N; i++) {
      if (latt->bond_list[latt->y_bonds[i]][1] == latt->plaquette_sites[p][5]) {
        latt->plaquette_bonds[p][5] = latt->y_bonds[i];
        break;
      }
    }
  }
}

void free_lattice(Lattice* latt) 
{
  int i, j;

  for (i = 0; i < latt->N; i++) {
    for (j = 0; j < latt->N; j++) {
      free(latt->r_ij[i][j]);
      free(latt->k_ij[i][j]);
    }
    free(latt->r_ij[i]);
    free(latt->k_ij[i]);
    free(latt->r[i]);
    free(latt->k[i]);
  }   
  free(latt->r);
  free(latt->r_ij);
  free(latt->k);

  for (i = 0; i < latt->Lx; i++) {
    free(latt->inv_r[i]);
    free(latt->inv_k[i]);
  }
  free(latt->inv_r);
  free(latt->inv_k);

  for (i = 0; i < latt->Nb; i++) {
    free(latt->bond_list[i]);
  }
  free(latt->bond_list);
  free(latt->r_sites);
  free(latt->sublattice);
  free(latt->bond_type);
  free(latt->x_bonds);
  free(latt->y_bonds);
  free(latt->z_bonds);

  for (i = 0; i < latt->Np; i++) {
    free(latt->plaquette_bonds[i]);
    free(latt->plaquette_sites[i]);
  }
  free(latt->plaquette_bonds);
  free(latt->plaquette_sites);
}

File: python_code/lattice.py
import numpy as np

class Lattice():
  # Real space unit cell vectors
  n1 = 0.5 * np.array([  1, np.sqrt(3)])
  n2 = 0.5 * np.array([- 1, np.sqrt(3)])
  # Reciprocal space unit cell vectors
  b1 = 2.0 * np.pi * np.array([n2[1], - n2[0]]) / (n1[0] * n2[1] - n1[1] * n2[0])
  b2 = 2.0 * np.pi * np.array([- n1[1], n1[0]]) / (n1[0] * n2[1] - n1[1] * n2[0])
  # Position of the lattice sites inside the unit cell
  __p1 = np.array([0.0, - 0.5])
  __p2 = np.array([0.0, 0.5])

  def __init__(self, Lx: int, Ly: int):
    self.Lx = Lx
    self.Ly = Ly
    self.N = self.Lx * self.Ly
    self.z = 3
    self.Ns = 2 * self.N
    self.Nb = self.z * self.Ns // 2
    
    # Real space and k space unit cell coordinates
    self.r_unit_cell = np.zeros((self.N, 2))
    self.inv_r_unit_cell = np.zeros((self.Lx, self.Ly), dtype=int)
    self.k_unit_cell = np.zeros((self.N, 2))
    self.inv_k_unit_cell = np.zeros((self.Lx, self.Ly), dtype=int)
   
    n = 0 
    for j in range(self.Ly):
      for i in range(self.Lx):
        self.r_unit_cell[n, :] = i * self.n1 + j * self.n2
        self.inv_r_unit_cell[i, j] = n
        self.k_unit_cell[n, :] = i * self.b1 / self.Lx + j * self.b2 / self.Ly
        self.inv_k_unit_cell[i, j] = n
        
        n += 1
   
    # Lattice site coordinates and bond matrix
    self.r_sites = np.zeros((self.Ns, 2))
    self.unit_cell = np.zeros((self.Ns), dtype=int)
    self.sublattice = np.zeros((self.Ns), dtype=int)
    self.bond_list = np.zeros((self.Nb, 2), dtype=int)
    self.bond_type = np.zeros((self.Nb), dtype=int)
    
    n = 0
    for i in range(self.Lx):
      for j in range(self.Ly):
        n_ = self.inv_r_unit_cell[i, j]
        
        self.r_sites[n, :] = self.r_unit_cell[n_, :] + self.__p1
        self.unit_cell[n] = n_
        self.sublattice[n] = 1
        n += 1
        
        self.r_sites[n, :] = self.r_unit_cell[n_, :] + self.__p2
        self.unit_cell[n] = n_
        self.sublattice[n] = 2
        n += 1

    nb = 0
    for i in range(self.Lx):
      for j in range(self.Ly):
        n = 2 * self.inv_r_unit_cell[i, j]
        
        # nn in z dir
        n1 = 2 * self.inv_r_unit_cell[i, j] + 1
        self.bond_list[nb, 0] = n
        self.bond_list[nb, 1] = n1
        self.bond_type[nb] = 3
        nb += 1
        
        # nn in y dir
        n1 = 2 * self.inv_r_unit_cell[i, (j + 1)%self.Ly] + 1
        self.bond_list[nb, 0] = n
        self.bond_list[nb, 1] = n1
        self.bond_type[nb] = 2
        nb += 1
        
        # nn in x dir
        n1 = 2 * self.inv_r_unit_cell[(i + 1)%self.Lx, j] + 1
        self.bond_list[nb, 0] = n
        self.bond_list[nb, 1] = n1
        self.bond_type[nb] = 1
        nb += 1
    
    # Brillouin zone for plotting
    gamma = 0.0 * self.b1 + 0.0 * self.b2
    gamma_prime = 1.0 * self.b1 + 0.0 * self.b2
    m1 = 0.5 * self.b1 + 0.5 * self.b2
    m3 = 0.5 * self.b1 + 0 * self.b2
    x = -0.5 * self.b1 + 0.5 * self.b2
    
    self.q_points = list()
 
    self.gamma_idx = 0 
    self.q_points.append(gamma)
    for n in range(self.N):
      if self.k_unit_cell[n, 0] == gamma[0] and self.k_unit_cell[n, 1] < m1[1] and self.k_unit_cell[n, 1] > gamma[1]:
        self.q_points.append(self.k_unit_cell[n, :])
    self.q_points.append(m1)
    self.m1_idx = len(self.q_points) - 1
    for n in range(self.N):
      if self.k_unit_cell[n, 1] == m1[1] and self.k_unit_cell[n, 0] > m1[0] and self.k_unit_cell[n, 0] < gamma_prime[0]:
        self.q_points.append(self.k_unit_cell[n, :])
    self.q_points.append(gamma_prime)
    self.gamma_prime_idx = len(self.q_points) - 1
    for i in range(self.Lx-1,-1,-1):
      n = self.inv_k_unit_cell[i, 0]
      if self.k_unit_cell[n, 1] > m3[1] and self.k_unit_cell[n, 0] > m3[0]:
        self.q_points.append(self.k_unit_cell[n, :])
    self.q_points.append(m3)
    self.m3_idx = len(self.q_points) - 1
    for i in range(self.Lx-1,-1,-1):
      n = self.inv_k_unit_cell[i, 0]
      if self.k_unit_cell[n, 1] < m3[1] and self.k_unit_cell[n, 0] < m3[0]:
        self.q_points.append(self.k_unit_cell[n, :])

    self.q_points = np.array(self.q_points)

    # Plaquettes 
    self.Np = self.N
    self.size_p = 6
    self.plaquettes_sites = np.zeros((self.Np, self.size_p), dtype=int)
    self.plaquettes_bonds = np.zeros((self.Np, self.size_p), dtype=int)
    
    self.z_bonds = list()
    self.y_bonds = list()
    self.x_bonds = list()
    p = 0
    for b in range(self.Nb):
      if self.bond_type[b] == 3:
        self.z_bonds.append(b)
      if self.bond_type[b] == 2:
        self.y_bonds.append(b)
      if self.bond_type[b] == 1:
        self.x_bonds.append(b)
    self.x_bonds = np.array(self.x_bonds)
    self.y_bonds = np.array(self.y_bonds)
    self.z_bonds = np.array(self.z_bonds)
    
    for p in range(self.Np):
      self.plaquettes_sites[p, 0] = self.bond_list[self.z_bonds[p], 0]
      self.plaquettes_sites[p, 1] = self.bond_list[self.z_bonds[p], 1]
      self.plaquettes_bonds[p, 0] = self.z_bonds[p]
      
      for i in range(self.x_bonds.shape[0]):
        if self.bond_list[self.x_bonds[i], 1] == self.plaquettes_sites[p, 1]:
          self.plaquettes_sites[p, 2] = self.bond_list[self.x_bonds[i], 0]
          self.plaquettes_bonds[p, 1] = self.x_bonds[i]
          break

      for i in range(self.y_bonds.shape[0]):
        if self.bond_list[self.y_bonds[i], 0] == self.plaquettes_sites[p, 2]:
          self.plaquettes_sites[p, 3] = self.bond_list[self.y_bonds[i], 1]
          self.plaquettes_bonds[p, 2] = self.y_bonds[i]
          break
        
      for i in range(self.z_bonds.shape[0]):
        if self.bond_list[self.z_bonds[i], 1] == self.plaquettes_sites[p, 3]:
          self.plaquettes_sites[p, 4] = self.bond_list[self.z_bonds[i], 0]
          self.plaquettes_bonds[p, 3] = self.z_bonds[i]
          break
        
      for i in range(self.x_bonds.shape[0]):
        if self.bond_list[self.x_bonds[i], 0] == self.plaquettes_sites[p, 4]:
          self.plaquettes_sites[p, 5] = self.bond_list[self.x_bonds[i], 1]
          self.plaquettes_bonds[p, 4] = self.x_bonds[i]
          break
        
      for i in range(self.y_bonds.shape[0]):
        if self.bond_list[self.y_bonds[i], 1] == self.plaquettes_sites[p, 5]:
          self.plaquettes_bonds[p, 5] = self.y_bonds[i]
          break
    

